be/src/formats/parquet/file_writer.cpp

```diff
@@ -177,136 +191,570 @@ void ParquetBuildHelper::build_compression_type(::parquet::WriterProperties::Bui
-    } else {                                                                                                 \
-        ::parquet::ByteArray value;                                                                          \
-        const BinaryColumn* binary_col = down_cast<const BinaryColumn*>(data_column);                        \
-        for (size_t i = 0; i < num_rows; i++) {                                                              \
-            value.ptr = reinterpret_cast<const uint8_t*>(binary_col->get_slice(i).get_data());               \
-            value.len = binary_col->get_slice(i).get_size();                                                 \
-            col_writer->WriteBatch(1, nullptr, nullptr, &value);                                             \
-        }                                                                                                    \
-    }                                                                                                        \
-    _buffered_values_estimate[i] = col_writer->EstimatedBufferedValueBytes();
+}
 
 Status FileWriterBase::write(Chunk* chunk) {
     if (!chunk->has_rows()) {
         return Status::OK();
     }
+    _generate_rg_writer();
+    _col_idx = 0; // reset index upon every write
 
-    Columns result_columns;
-    // Step 1: compute expr
-    int num_columns = _output_expr_ctxs.size();
-    result_columns.reserve(num_columns);
-
-    for (int i = 0; i < num_columns; ++i) {
-        ASSIGN_OR_RETURN(ColumnPtr column, _output_expr_ctxs[i]->evaluate(chunk));
-        //column = _output_expr_ctxs[i]->root()->type().type == TYPE_TIME
-        //         ? vectorized::ColumnHelper::convert_time_column_from_double_to_str(column)
-        //         : column;
-        result_columns.emplace_back(std::move(column));
-    }
-
-    size_t num_rows = chunk->num_rows();
-    for (size_t i = 0; i < result_columns.size(); i++) {
-        // auto &col = chunk->get_column_by_index(i);
-        auto& col = result_columns[i];
-        bool nullable = col->is_nullable();
-        auto null_column = nullable && down_cast<starrocks::NullableColumn*>(col.get())->has_null()
-                                   ? down_cast<starrocks::NullableColumn*>(col.get())->null_column()
-                                   : nullptr;
-        const auto data_column = ColumnHelper::get_data_column(col.get());
-
-        std::vector<int16_t> def_level(num_rows);
-        std::fill(def_level.begin(), def_level.end(), 1);
-        if (null_column != nullptr) {
-            auto nulls = null_column->get_data();
-            for (size_t j = 0; j < num_rows; j++) {
-                def_level[j] = nulls[j] == 0;
-            }
+    Context ctx(0, chunk->num_rows());
+    for (int i = 0; i < chunk->num_rows(); i++) {
+        ctx.append(i, 0, 0);
+    }
+
+    for (size_t i = 0; i < chunk->num_columns(); i++) {
+        auto col = chunk->get_column_by_index(i);
```

Comment:
ASSIGN_OR_RETURN(ColumnPtr column, _output_expr_ctxs[i]->evaluate(chunk));

Reply:
This can cause errors

Reply:
I wonder in which cases `chunk->get_column_by_index(i)` causes error